compiler/rustc_hir_typeck/src/op.rs

```diff
@@ -198,6 +198,29 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    fn check_expr_coercible_to_type_or_error(
+        &self,
+        expr: &'tcx hir::Expr<'tcx>,
+        expected: Ty<'tcx>,
+        expected_ty_expr: &'tcx hir::Expr<'tcx>,
+        op: hir::BinOp,
+    ) -> Ty<'tcx> {
+        let ty = self.check_expr_with_hint(expr, expected);
+        // checks don't need two phase
+        match self.demand_coerce_diag(expr, ty, expected, Some(expected_ty_expr), AllowTwoPhase::No)
+        {
+            Ok(ty) => ty,
+            Err(mut err) => {
+                self.suggest_swapping_lhs_and_rhs(&mut err, ty, expected_ty_expr, op);
```

Comment:
use the `extend_err: impl FnOnce(&mut Diag<'_>)` argument as with `check_expr_has_type_or_error`. Then you can make `check_expr_coercible_to_type` call `check_expr_coercible_to_type_or_error` with an empty closure to prevent duplication.